components/omega/src/base/MachEnv.h

```diff
@@ -0,0 +1,166 @@
+#ifndef OMEGA_MACH_ENV_H
+#define OMEGA_MACH_ENV_H
+//===-- base/MachEnv.h - machine environment definitions --------*- C++ -*-===//
+//
+/// \file
+/// \brief Defines aspects of the parallel and node machine environment
+///
+/// The machine environment defines a number of parameters associated with
+/// the message-passing and threading environments. It also can describe
+/// the node-level hardware environment, including any useful accelerator
+/// or processor-level constants. Multiple machine environments can be defined,
+/// mostly associated with subsets of the processor decomposition, but
+/// a default machine env must be created very early in model
+/// initialization. All environments can be retrieved by name, though a
+/// specific retrieval for the most common default environment is provided
+/// for efficiency.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mpi.h"
+
+#include <map>
+#include <string>
+
+namespace OMEGA {
+
+// For CPUs, a compile-time vector length is useful for
+// blocking the inner loops. This should be set to an appropriate
+// length (typically 32, 64, 128) for CPU-only builds, but set to one for
+// GPU builds to maximize parallelism instead.
+#ifdef OMEGA_VECTOR_LENGTH
+constexpr int VecLength = OMEGA_VECTOR_LENGTH;
+#else
+constexpr int VecLength = 1;
```

Comment:
Good to think about, but we can probably hold off on that level of granular configuration for GPUs until needed.

Reply:
Since this vector length will often appear as vector blocking of the inner loop, we will generally want it to be 1 for GPU and a separate GPU_VECTOR_LENGTH would be redundant. (Basically like the pack they're doing in the atmosphere). Though I think Trey was exploring use of non-unity values to match warps/thread group sizes and we could set this vector length to an appropriate value in that case. I do think we eventually want some GPU-related optimizing parameters, but Brian is correct that we want to figure out what those should be and can add later.